// @tsed/mongoose/lib/decorators/dynamicRef.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@tsed/common");
const core_1 = require("@tsed/core");
const mongoose_1 = require("mongoose");
const constants_1 = require("../constants");
/**
 * Define a property as mongoose reference to other Model (decorated with @Model).
 *
 * ### Example
 *
 * ```typescript
 * @Model()
 * class FooModel {
 *
 *    @DynamicRef('type')
 *    field: DynamicRef<OtherFooModel | OtherModel>
 *
 *    @Enum(['OtherFooModel', 'OtherModel'])
 *    type: string
 * }
 *
 * @Model()
 * class OtherFooModel {
 * }
 *
 * @Model()
 * class OtherModel {
 * }
 * ```
 *
 * @param refPath
 * @returns {Function}
 * @decorator
 * @mongoose
 * @property
 */
function DynamicRef(refPath) {
    return core_1.applyDecorators(common_1.Property({ use: String }), common_1.Schema({
        type: String,
        example: "5ce7ad3028890bd71749d477",
        description: "Mongoose Ref ObjectId"
    }), core_1.StoreFn((store) => {
        delete store.get("schema").$ref;
    }), core_1.StoreMerge(constants_1.MONGOOSE_SCHEMA, {
        type: mongoose_1.Schema.Types.ObjectId,
        refPath
    }));
}
exports.DynamicRef = DynamicRef;
